src/transformers/utils/import_utils.py

```diff
@@ -1581,3 +1655,173 @@ def direct_transformers_import(path: str, file="__init__.py") -> ModuleType:
     spec.loader.exec_module(module)
     module = sys.modules[name]
     return module
+
+
+def register(*, backends=()):
+    """
+    This method enables two things:
+    - Attaching a `__backends` tuple to an object to see what are the necessary backends for it
+      to execute correctly without instantiating it
+    - The '@register' string is used to dynamically import objects
+    """
+
+    if not isinstance(backends, tuple):
+        raise ValueError("Backends should be a tuple.")
+
+    def inner_fn(fun):
+        fun.__backends = backends
+        return fun
+
+    return inner_fn
+
+
+@lru_cache()
+def define_import_structure(module_path):
+    import_structure = {}
+    if os.path.isdir(module_path):
+        for f in os.listdir(module_path):
+            if f != "__pycache__" and os.path.isdir(os.path.join(module_path, f)):
+                import_structure[f] = define_import_structure(os.path.join(module_path, f))
+        directory = module_path
+    else:
+        directory = os.path.dirname(module_path)
+
+    adjacent_modules = [f for f in os.listdir(directory) if not os.path.isdir(os.path.join(directory, f))]
```

Comment:
2 things here:
- in case `module_path` is a directory (checked above), we are doing `os.listdir(directory)` twice.
- in case `module_path` is a file, we are doing `os.listdir(directory)` once per file in the folder (IIUC, since we do `define_import_structure(os.path.join(module_path, f))` on all files from the same dir) 

Since `os.listdir` might be an heavy operation, worth checking if it can be optimized (or at least cached with a lru_cache?) to save time.

Reply:
I changed how that behaved so it'll be more efficient in 5d76ab742e